scikits/odes/sundials/cvode.pyx

```diff
@@ -1155,27 +1272,61 @@ cdef class CVODE:
         cdef N_Vector y  = self.y
 
         y_last   = np.empty(np.shape(y0), float)
+        t = tspan[idx]
 
-        for idx in np.arange(np.alen(tspan))[1:]:
-            t = tspan[idx]
-
+        while True:
             flag = CVode(cv_mem, <realtype> t,  y, &t_out, CV_NORMAL)
 
             nv_s2ndarray(y,  y_last)
 
-            if flag != CV_SUCCESS:
-                # return values computed so far
-                t_retn  = t_retn[0:idx]
-                y_retn  = y_retn[0:idx, :]
+            if flag == CV_SUCCESS or flag == CV_WARNING:
+                t_retn[idx]    = t_out
+                y_retn[idx, :] = y_last
+                idx = idx + 1
+                PyErr_CheckSignals()
+
+                # Iterate until we reach the end of tspan
+                if idx < last_idx:
+                    t = tspan[idx]
+                    continue
+                else:
+                    return (
+                        flag, t_retn, y_retn, None, None, t_roots, y_roots,
+                        None, None, None
+                    )
 
-                return flag, t_retn, y_retn, t_out, y_last
+            elif flag == CV_ROOT_RETURN:
+                t_roots.append(np.copy(t_out))
+                y_roots.append(np.copy(y_last))
+                break
+            elif flag == CV_TSTOP_RETURN:
+                t_tstop.append(np.copy(t_out))
+                y_tstop.append(np.copy(y_last))
+                break
 
-            t_retn[idx]    = t_out
-            y_retn[idx, :] = y_last
+            break
+
+
+        PyErr_CheckSignals()
 
-            PyErr_CheckSignals()
+        # return values computed so far
+        t_retn  = t_retn[0:idx]
+        y_retn  = y_retn[0:idx, :]
+        if flag == CV_ROOT_RETURN:
+            return (
+                flag, t_retn, y_retn, None, None, t_roots, y_roots,
+                t_tstop, y_tstop,
+            )
+        elif flag == CV_TSTOP_RETURN:
+            return (
+                flag, t_retn, y_retn, None, None, t_roots, y_roots,
+                t_tstop, y_tstop,
+            )
+        return (
+            flag, t_retn, y_retn, t_out, y_last, t_roots, y_roots,
+            t_tstop, y_tstop,
+        )
```

Comment:
I have a feeling this can all be collapsed to a single return, provided we check empty [] and replace by None. 
On the other hand, lists would be nicer as numpy arrays, but then, these lists t_stop/roots and y-stop/roots never contain more than one entry it seems.